src/protocol/malicious.rs

```diff
@@ -0,0 +1,350 @@
+        }
+    }
+}
+
+#[allow(dead_code)]
+pub struct SecurityValidator<'a, F, N> {
+    ctx: ProtocolContext<'a, N>,
+    r_share: Replicated<F>,
+    u: F,
+    w: F,
+}
+
+impl<'a, F: Field, N: Network> SecurityValidator<'a, F, N> {
+    #[allow(dead_code)]
+    pub fn new(ctx: ProtocolContext<'a, N>) -> SecurityValidator<'a, F, N> {
+        let prss = ctx.prss();
+
+        let r_share = prss.generate_replicated(RecordId::from(0));
+        let (u_left, u_right): (F, F) = prss.generate_fields(RecordId::from(1));
+        let (w_left, w_right): (F, F) = prss.generate_fields(RecordId::from(2));
+
+        SecurityValidator {
+            ctx,
+            r_share,
+            u: u_right - u_left,
+            w: w_right - w_left,
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn r_share(&self) -> Replicated<F> {
+        self.r_share
+    }
+
+    #[allow(dead_code)]
+    fn compute_dot_product_contribution(a: Replicated<F>, b: Replicated<F>) -> F {
+        (a.left() + a.right()) * (b.left() + b.right()) - a.right() * b.right()
+    }
+
+    #[allow(dead_code)]
+    #[allow(clippy::needless_pass_by_value)]
+    pub fn accumulate_macs(
+        &mut self,
+        ctx: ProtocolContext<'_, N>,
+        record_id: RecordId,
+        x: Replicated<F>,
+        rx: Replicated<F>,
+    ) {
+        // The helpers need to use the same shared randomness to generate the random constant used to validate a given multiplication.
+        // This is a bit tricky, because we cannot count on the multiplications being executed in the same order across all the helpers.
+        // The easiest way is to just narrow the context used to perform the multiplication, and then re-use the same record_id.
+        // This ensures that when the helpers all go to validate the multiplication: "1/foo/bar/baz", they all use the prss from "1/foo/bar/baz/validate".
+        // That way, we don't need to worry about the order in which the multiplications are executed.
+        let random_constant = ctx.prss().generate_replicated(record_id);
+
+        self.u += Self::compute_dot_product_contribution(random_constant, rx);
+        self.w += Self::compute_dot_product_contribution(random_constant, x);
+    }
+
+    #[allow(dead_code)]
+    pub async fn validate(&self) -> Result<(), BoxError> {
+        let record_3 = RecordId::from(3);
+        let record_4 = RecordId::from(4);
+        let record_5 = RecordId::from(5);
+        let record_6 = RecordId::from(6);
+
+        // send our `u_i+1` value to the helper on the right
+        let channel = self.ctx.mesh();
+        try_join(
+            channel.send(
+                self.ctx.role().peer(Direction::Right),
+                record_3,
+                UValue { payload: self.u },
+            ),
+            channel.send(
+                self.ctx.role().peer(Direction::Right),
+                record_4,
+                UValue { payload: self.w },
+            ),
+        )
+        .await?;
```

Comment:
I want to flag this await here.  You don't use the value, nor do you depend on this operation completing in any way.  However, I believe that this is the correct approach anyway, because this is how we will ultimately manage back pressure.  Inability to send will then block further processing at this point.

This is not obvious and there is a risk that someone decides to get smart and defer the await on this promise until the very end.

The same does not apply to the use of `try_join()` here, though I think that you could just do two `channel.send(...).await?` calls in sequence.  The two records are being sent on the same channel, so you don't need to allow for them arriving out of order.